### ISSUE TITLE
NVDA installation may fail with an error if system UI language and language for non-unicode programs don't match

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#5166]
**Reported by k_kolev1985 on 2015-06-20 07:56**
Hello,

The tests are done on 2 machines with Windows 7 Ultimate 32-bit, but I think that the issue will occur on every version of Windows supported by NVDA, if the conditions are met. In both cases, the UI language was set to "bulgarian", but the language for non-unicode programs was set to something else, witch doesn't support cyrillic (bulgarian) text ("English (United States)" is such a language). In such a case, NVDA will display an error message dialog with the information that the installation failed and for more details to check the NVDA log file. The log file reports an error related to the creation of the NVDA desktop shortcut (see the attached log files). However, the shortcut seams to be created OK - it works as it should, including the Control+Alt+N keyboard shortcut to launch or restart NVDA (it is set in the shortcut's properties). If the UI language and the language for non-unicode programs match, there is no issue. I suspect that the issue will occur only if the language for non-unicode programs set doesn't support displaying of text in the UI language set.

BTW: As far as I know, it is not a common thing for the UI language and the language for non-unicode programs to mismatch on a standard PC configuration (at least the language for non-unicode programs is set correctly). But sometimes (such as in the case mentioned above), it may be set so by mistake. In normal circumstances, we here in Bulgaria set the language for non-unicode programs to "Bulgarian", so we don't have issues with displaying cyrillic text in non-unicode programs.

Steps to reproduce the issue:
1. In Control Panel -> Region and Language, set the UI language and the language for non-unicode programs to such values, so the language for non-unicode programs will not support displaying of text in the UI language set. Reboot if required.
2. I'm not sure that it is required, but just to be sure - delete the NVDA shortcut on the desktop.
3. Start installing or updating NVDA from a downloaded installer.
4. In the installer, check the option to create a shortcut for NVDA on the desktop.
5. Wait for the installer to do its job.

Actual results:
NVDA will display an error message dialog with the information that the installation failed and for more details to check the NVDA log file.

Expected results:
The NVDA installation/update should complete with no such errors, even if the UI language and the language for non-unicode programs don't match.

Test environment:
- Operating system: Windows 7 Ultimate 32-bit.
- NVDA version: 2015.2.

**Comment by Adriani90** [comment]
@k-kolev1985 I know that this might be difficult to test, but do you think this issue is still occuring? cc: @josephsl

**Comment by k-kolev1985** [comment]
@Adriani90 Since there hasn't been any work done in order to fix this in NVDA, I doubt that is fixed now. I may be able to test it in one of my virtual machines these days and report the results here. I just need to find some free time to do it. :-)

**Comment by k-kolev1985** [comment]
@Adriani90 Yes, the bug is still reproducible with latest NVDA alpha snapshot and Windows 10 build 18323 Insider Preview, as described in the first post here. But if I check the beta option to use Unicode UTF-8 for locale support in Windows (under the combobox for selecting the language for non-Unicode programs), the problem does not occur, even if the display/UI language and the language for non-Unicode programs don't match.

**Comment by leonardder** [comment]
I assume we can somehow fix this when switching to Python 3.

**Comment by k-kolev1985** [comment]
I've tested with NVDA snapshot alpha 18450,8c27752f and the error during install still occurred. It wasn't on a clean install of NVDA, but I think that does not matter. The testing is done on Windows 10 (build 18963 I think), 32-bit, in bulgarian, with language for non-Unicode programs set to "English U.S.".

**Comment by leonardder** [comment]
"Could you please provide a log for the alpha build?

**Comment by k-kolev1985** [comment]
OK, here is the log from the failed installed of the alpha build.
[NVDA_Log_2019-08-27_#5166.log](https://github.com/nvaccess/nvda/files/3547466/NVDA_Log_2019-08-27_.5166.log)

**Comment by leonardder** [comment]
This looks like an issue in the WshShell interfaces. When I'm not using dynamic dispatch, the issue seems to be that the path is silently converted to mbcs, any unicode characters get lost and are replaced by a question mark. However, question marks are not allowed in file names, and therefore neither in shortcut names.
We could solved this by encoding to mbcs with errors="ignore" and than decoding back, however, this will cause anything that's not accepted by mbcs to get lost in the process.

@k-kolev1985 Couldn't you set the language for non-unicode programs to Bulgarian instead, or does that cause other issues for you?

**Comment by k-kolev1985** [comment]
@leonardder I don't do that change intentionally. Neither are the other bulgarian Windows users. We know that we should set the language for non-unicode programs to "Bulgarian" for best results. But by mistake it may happen. Once, I was configuring a new virtual machine with Windows 10 and I forgot to set the language for non-unicode programs to "Bulgarian" and had to deal with problems with text in cyrillic not displaying correctly in non-unicode programs and text files, before I realized my mistake. Recently it happened to a friend of mine - he forgot to set the language for non-unicode programs to "Bulgarian" and got the same error during install of NVDA. So, such configuration is not intentional, but by mistake. I don't know of cases in which is/was required to set the language for non-unicode programs to something other than "Bulgarian" for some purposes, but there may be such cases - it wouldn't surprise me.

**Comment by jcsteh** [comment]
We've seen this problem before in #6326. Although this was reported as fixed in https://github.com/nvaccess/nvda/issues/6326#issuecomment-334933963, I don't think the underlying problem was really fixed. Perhaps the localisation worked around it or the user changed system settings.

I did a pretty in-depth analysis of this and provided a possible solution in https://github.com/nvaccess/nvda/issues/6326#issuecomment-326152007 . Including it below since that issue is now closed.

> It looks like this happens for any character that can't be encoded using the system ANSI code page:
> 
> ```
> >>> u"\u2022".encode("mbcs")
> '\x95'
> >>> u"ț".encode("mbcs")
> '?'
> >>> u"‣".encode("mbcs")
> '?'
> ```
> 
> If I try creating the shortcut with either of the latter two, I get an exception.
> 
> Even though we pass the string as Unicode, wScript.Shell seems to convert it to ANSI. It seems [others have hit this problem too](https://superuser.com/questions/1129549/ascii-in-powershell-wscript-shell). Note that this is definitely not a file system issue; you can quite happily write a file containing these characters if you use native Unicode:
> 
> ```
> >>> file(u"c:\\users\\jamie\\junk\\jtd‣bar", "w")
> <open file u'c:\\users\\jamie\\junk\\jtd\u2023bar', mode 'w' at 0x070573E8>
> ```
> 
> I wonder whether we'll need to switch to an alternative way of creating these shortcuts. We should be able to do this using [IShellLink](https://msdn.microsoft.com/en-us/library/bb774950(v=vs.85).aspx) and friends. IShellLink specifically mentions IShellLinkA and IShellLinkW interfaces, which suggests the latter does explicitly do Unicode. See [this article](https://msdn.microsoft.com/en-us/library/windows/desktop/bb776891(v=vs.85).aspx) for further details. I did a bit of investigation on this in the labyrinth that is MSDN:
> 
>     1. I found CLSID_ShellLink in ShObjIdl.idl (00021401-0000-0000-C000-000000000046). Unfortunately, it looks like there's no typelib registered for this.
> 
>     2. There are also dispatch interfaces for most (all?) of this stuff. You access them by using the "shell.application" progId. See [Scriptable Shell Objects](https://msdn.microsoft.com/en-us/library/windows/desktop/bb776890(v=vs.85).aspx) and [ShellLinkObject.Save](https://msdn.microsoft.com/en-us/library/bb773998(v=vs.85).aspx) for details/examples.
